### ISSUE TITLE
[RFC] Export keyword to indicate the results of a query.

### ISSUE THREAD

**Issue by nathanielc** [influxdata/ifql#143]
Once an IFQL query has more than one statement, it will be required to define which statements results should be returned to the client.

Using an `export` keyword will indicate what steps of the pipeline should be returned.

Simple example exporting all results
```
export cpu = from(db:"mydb").filter(exp:"_measurement" == "cpu")
export mem = from(db:"mydb").filter(exp:"_measurement" == "mem")
```

Exports are named so that the client knows which results are associated with which exports.

An export can also get its name from a var.

```
var cpu = from(db:"mydb").filter(exp:"_measurement" == "cpu")

export cpu
```

Exports will only be allowed when running queries as adhoc queries. Otherwise the query will error as there will not be a client to receive the data.

Some notes:

* The `export` keywords feels like the opposite of the `import` keyword, but it is not, so we need to choose a better name. We considered the name `return` but that reads like it should be the end of the script which isn't the case.

**Comment by nathanielc** [comment]
What about the keyword `result`? The doesn't read like `return` but still indicates that the expression is the result of the query.

**Comment by nathanielc** [comment]
Hmm, but `result` is a noun and it reads better if its a verb.

Maybe one of these:

* reply
* respond
* dump
* answer
* yield (this would probably be confusing as its not semantically the same as a more traditional yield)

**Comment by rbetts** [comment]
Beam seems to use `output`:  `output.apply(TextIO.write().to("gs://some/outputData"));` (https://beam.apache.org/documentation/programming-guide/#pipeline-io)

Flink uses `sinks`: https://ci.apache.org/projects/flink/flink-docs-release-1.3/dev/datastream_api.html#data-sinks

Generally, I'm not quite sure about a keyword that implies a single sink. For example, in a streaming mode, it would be common to sink results to one output and errors to another.

**Comment by desa** [comment]
@nathanielc what about changing the name of `import`, Maybe something like `load`.

**Comment by desa** [comment]
* send
* push
* respond
* select
* output
* fetch
* feed
* marshal
* stream

**Comment by desa** [comment]
Another possibility, instead of `return` maybe add an `ing` at the end of it.

So it'd end up being

```
returning cpu = from(db:"mydb").filter(exp:"_measurement" == "cpu")
returning mem = from(db:"mydb").filter(exp:"_measurement" == "mem")
```

I think that makes it more obvious that a `return-ing` isn't terminal.

**Comment by nathanielc** [comment]
After some discussion, we like dropping the idea of the keyword and instead using global functions for all outputs.

Something like this:

```javascript
yield(name:"cpu", data: from(db:"mydb").filter(exp:"_measurement" == "cpu"))
yield(name:"mem", data: from(db:"mydb").filter(exp:"_measurement" == "mem"))
```

This works well for when you are streaming and want to explicitly send data to different endpoints.



```javascript
// Send data to the client
yield(name:"cpu", data: from(db:"mydb").filter(exp:"_measurement" == "cpu"))

 // Write data back to influx
influx(db:"newdb", data: from(db:"mydb").filter(exp:"_measurement" == "mem"))
```

**Comment by rbetts** [comment]
@nathanielc So essentially `yield` means `select`?  Is `yield` reserved for results returned to an ad-hoc client?  I'm not quite sure I follow the conclusion.

**Comment by nathanielc** [comment]
@rbetts Yes and yes, `yield` means send the data to the client and as such will not make sense in a streaming based query, only for ad-hoc queries. If a `yield` function is used in a non ad-hoc query than an error will be returned.

And it still stands that if your query only has a single statement there is no need to explicitly call `yield` as its clear that the client wants the result of the statement. A default name will be chosen, but it doesn't really matter what the name is since it doesn't need to distinguish from other results.

**Comment by rbetts** [comment]
@nathanielc I think a lot of users will want to use adhoc to test an eventual streaming or batch query with minimal edits once they like the adhoc result. For batch, you might test a subset of the batch. And even ad-hoc results will stream (when possible) to the client. 

A few questions as a potential user: 
 - Does the first`yield` return all results before returning results from a second `yield`?
 - Does each `yield` require its own filter? Or can I `yield` for example `min()` and `max()` from a `filter`?
 - Why is `yield` a verb but `influx` is a noun in the example?
 - If I want to run an adhoc query to see the output of `select .. into` and then run my tested query to write a downsample back to the database, how much IFQL code do I have to change? (Ideally the answer here is zero.)

**Comment by nathanielc** [comment]
> Does the firstyield return all results before returning results from a second yield?

No, the results can be intermixed. The will be delivered in the order they are produced from the query engine. The name of the yield is used to distinguish them.

> Does each yield require its own filter? Or can I yield for example min() and max() from a filter?

Yield makes no restrictions on the `data` parameter, it need only be a table(i.e. the result of any function chain).

> Why is yield a verb but influx is a noun in the example?

Influx should have a better name, maybe `writeInflux` or simply `write`, since we are in the context of influx already.

> If I want to run an adhoc query to see the output of select .. into and then run my tested query to write a downsample back to the database, how much IFQL code do I have to change? (Ideally the answer here is zero.)

Hmm this is an interesting one, I agree that the transition from "I am running an ad-hoc query for testing" and "the query is complete so I want to run it for realz" needs to be easy. I guess I have a question in return, is it possible to do a `select ... into` as an ad-hoc query that actually writes data back to the database? Or in other words what is the difference between and ad-hoc query and a streaming query? In my mind the difference is that an ad-hoc query is bounded in time and is only run once and there is a client that can receive results. Where as a streaming query is not necessarily bound in time and can run multiple times (or even continuously) and there is no client that can receive results, (in other words all results must be actions against other systems like trigger an alert or write data back into a database).

Using these definitions it is possible to submit an ad-hoc query that writes data back into the database. In that case the client probably doesn't get any response other than successful completion of the query. So if it possible to write results back to a database in an ad-hoc query then how do you distinguish between writing results to a database and returning them to a client? I think we need two different functions as they are different actions.

Here is an imaginary workflow:

The goal is to construct a query that writes some downsampling result into an influxdb database, but we want to verify the results before writing any data.

1. Write a query the yields the results to the client

```javascript
from(db:"telegraf")
    .range(start:-1h)
    .where(exp:{"_measurement" == "requests" and "_field" == "count"})
    .window(every:5m)
    .mean()
```
> NOTE: The yield is not needed here as we only have  a single statement
2. Verify that the client got the expected results
3. Decide that instead of the mean request count you want the mean of the request count/error ratio.
4. Update the query as needed.
```javascript
var count = from(db:"telegraf")
    .range(start:-1h)
    .where(exp:{"_measurement" == "requests" and "_field" == "count"})

var errors = from(db:"telegraf")
    .range(start:-1h)
    .where(exp:{"_measurement" == "requests" and "_field" == "errors"})

var error_rate = join(on:["service"], eval: {errors/count})
    .window(every:5m)
    .mean()

yield(name: "error_rate", data:error_rate)
```
> NOTE: Now a yield is necessary since the query is more complex we want the return value to be explicit.

5a. Satisfied that the query is working they way we want it, run the query once in ad-hoc mode, replacing `yield` with `write` and the required db parameter.

```javascript
var count = from(db:"telegraf")
    .range(start:-1h)
    .where(exp:{"_measurement" == "requests" and "_field" == "count"})

var errors = from(db:"telegraf")
    .range(start:-1h)
    .where(exp:{"_measurement" == "requests" and "_field" == "errors"})

var error_rate = join(on:["service"], eval: {errors/count})
    .window(every:5m)
    .mean()

write(db: "error_rate", data:error_rate)
```

5b. Satisfied that the query is working they way we want it, run the query contiously forever until we tell it to stop (i.e. in streaming mode), replacing `yield` with `write` and the required db parameter. Also remove the range statements as we are now running an unbounded query.

```javascript
var count = from(db:"telegraf")
    .where(exp:{"_measurement" == "requests" and "_field" == "count"})

var errors = from(db:"telegraf")
    .where(exp:{"_measurement" == "requests" and "_field" == "errors"})

var error_rate = join(on:["service"], eval: {errors/count})
    .window(every:5m)
    .mean()

write(db: "error_rate", data:error_rate))
```
Notice the the ad-hoc and streaming query are the same for the `write` statement. This is because the distinction between ad-hoc and streaming is about the time bounds and how many times the query runs.
In this case the ad-hoc query is time bounded to be data between [-1h, now), but the streaming query is not time bounded at all. 

In short the difference between `yield` and `write` is about what you want to do with the data as opposed to how the query is running.


The differences between the two modes of query operation need to be explored in more detail this is a first pass at some of my ideas.

**Comment by desa** [comment]
@nathanielc I've been thinking about this as well. I've gone back and forth between the solution you suggest above, which is essentially replace any yields with whatever you'd like them to be.

Another alternative I've been thinking of is what if there is only a single top level function `yield` and there are "yield handler" that specify how the data from the `yield` should be handled. 

When running the query in ad-hoc mode, the default behavior of any yield would be to return the values back to the user.

So the user would specify
```js
var count = from(db:"telegraf")
    .range(start:-1h)
    .where(exp:{"_measurement" == "requests" and "_field" == "count"})

var errors = from(db:"telegraf")
    .range(start:-1h)
    .where(exp:{"_measurement" == "requests" and "_field" == "errors"})

var error_rate = join(on:["service"], eval: {errors/count})
    .window(every:5m)
    .mean()

yield(name: "error_rate", data:error_rate)
```
for ad-hoc queries and when they'd like to switch into streaming mode they'd provide additional parameters (kind of like a templated task).

```json
{
"yield-name": "error_rate",
"type": "writeInfluxDB",
"options": {
     "db": "mydb",
  },
}
```

A more complicated example would be

```js
var count = from(db:"telegraf")
    .range(start:-1h)
    .where(exp:{"_measurement" == "requests" and "_field" == "count"})

var errors = from(db:"telegraf")
    .range(start:-1h)
    .where(exp:{"_measurement" == "requests" and "_field" == "errors"})

var error_rate = join(on:["service"], eval: {errors/count})

yield(name: "error_rate", data: error_rate)
    .writeInflux(db: "error_rate")

var avg_error_rate = error_rate
    .window(every:5m)
    .mean()

yield(name: "other", data: avg_error_rate)
```

When this query is ran in ad-hoc mode, the user would receive the result of `avg_error_rate`. When they wanted to switch into streaming mode, they'd have to supply some additional information about what to do with the `"other"` yield data.

Hopefully my high level idea comes across. I think I may not be communicating the concept very well.

Some things I like about this approach are

1. There are 3 top level functions: `from` from reading  data, `join` for merging data, and `yield` for outputting data.
2. A single IFQL file is can sufficient to describe an entire streaming query if the user is explicit.

The one thing I'm not a huge fan of is having to be explicit about yield-handlers if I haven't specified them when I define a streaming query.

**Comment by rbetts** [comment]
> What is the difference between an ad-hoc query and a streaming query

An ad-hoc query runs against a finite data set. A streaming query runs against an unbounded data set. 

> remove the range query to get an unbounded query

How would I express "a moving average of all data between `now()` and `now() - 1h`?  Dropping `range(start:-1h)` to change from adhoc to streaming surprises me.

> re: @desa's proposal around metadata

I think the metadata you are proposing is similar to an `execute` keyword that runs the same `ifql` logic with different contexts: streaming | ad-hoc,  low-priority | high-priority, ...
